vowpalwabbit/learner.cc

```diff
@@ -32,32 +32,136 @@ const char* to_string(prediction_type_t prediction_type)
 
 namespace LEARNER
 {
+
+void dispatch_end_pass(vw& all, example* ec)
+{
+  all.l->end_pass();
+  VW::finish_example(all, ec);
+}
+
+void save(vw& all, example* ec)
+{
+  // save state command
+  string final_regressor_name = all.final_regressor_name;
+
+  if ((ec->tag).size() >= 6 && (ec->tag)[4] == '_')
+    final_regressor_name = string(ec->tag.begin()+5, (ec->tag).size()-5);
+
+  if (!all.quiet)
+    all.opts_n_args.trace_message << "saving regressor to " << final_regressor_name << endl;
+  save_predictor(all, final_regressor_name, 0);
+
+  VW::finish_example(all,ec);
+}
+
+bool inline is_save_cmd(example* ec)
+{
+  return (ec->tag.size() >= 4) && (0 == strncmp((const char*)ec->tag.begin(), "save", 4));
+}
+
 void process_example(vw& all, example* ec)
 {
   if (ec->indices.size() > 1) // 1+ nonconstant feature. (most common case first)
     dispatch_example(all, *ec);
   else if (ec->end_pass)
+    dispatch_end_pass(all, ec);
+  else if (is_save_cmd(ec))
+    save(all, ec);
+  else // empty example
+    dispatch_example(all, *ec);
+}
+
+void process_multi_ex(vw& all, multi_ex& ec_seq)
+{
+  all.learn(ec_seq);
+  all.l->finish_example(all, ec_seq);
+}
+
+bool example_is_test(example& ec, vw& all)
+{
+  return all.l->is_test_example(ec);
+}
+
+/* example headers have the word "shared" */
+bool ec_is_example_header(example& ec)  
+{
+  v_array<CB::cb_class> costs = ec.l.cb.costs;
+  if (costs.size() != 1) return false;
+  if (costs[0].probability == -1.f) return true;
+  return false;
+}
+
+/* is this just a newline */
+inline bool example_is_newline_not_header(example& ec)
+{ return (example_is_newline(ec) && !ec_is_example_header(ec));}
+
+ /* Adds an example to multiline collection
+  * Returns: true if complete and false if incomplete example */
+bool complete_multi_ex(example* ec, multi_ex& ec_seq, vw& all)
+{
+  const bool is_test_ec = example_is_test(*ec,all);
+  const bool need_to_break = VW::is_ring_example(all, ec) && (ec_seq.size() >= all.p->ring_size - 2);
+
+  if ((example_is_newline_not_header(*ec) && is_test_ec) 
+      || need_to_break)
   {
-    all.l->end_pass();
-    VW::finish_example(all, ec);
+    ec_seq.push_back(ec);
```

Comment:
Why add the newline example?  It seems like asking for trouble to double delimit the end of a multiline example?